wled00/wled.cpp

```diff
@@ -697,7 +701,48 @@ bool WLED::initEthernet()
 #else
   return false; // Ethernet not enabled for build
 #endif
+}
+
+// performs asynchronous scan for available networks (which may take couple of seconds to finish)
+// returns true if only one wifi is configured or scan completed
+bool WLED::findWiFi(bool doScan) {
+  if (multiWiFi.size() <= 1) {
+    DEBUG_PRINTLN(F("Defaulf WiFi used."));
+    selectedWiFi = 0;
+    return true;
+  }
+
+  if (doScan) WiFi.scanDelete();  // restart scan
+
+  int status = WiFi.scanComplete();
 
+  if (status == WIFI_SCAN_FAILED) {
+    DEBUG_PRINTLN(F("WiFi scan started."));
+    WiFi.scanNetworks(true);  // start scanning in asynchronous mode
+    return false;
+  }
+  if (status >= 0) { // status contains number of found networks
+    DEBUG_PRINT(F("WiFi scan completed: ")); DEBUG_PRINTLN(status);
+    int rssi = -9999;
+    for (int o = 0; o < status; o++) {
+      DEBUG_PRINT(F(" WiFi available: ")); DEBUG_PRINT(WiFi.SSID(o));
+      DEBUG_PRINT(F(" RSSI: ")); DEBUG_PRINT(WiFi.RSSI(o)); DEBUG_PRINTLN(F("dB"));
+      for (unsigned n = 0; n < multiWiFi.size(); n++)
+        if (!strcmp(WiFi.SSID(o).c_str(), multiWiFi[n].clientSSID)) {
+          // find the WiFi with the strongest signal (but keep priority of entry if signal difference is not big)
+          if ((n < selectedWiFi && WiFi.RSSI(o) > rssi-10) || WiFi.RSSI(o) > rssi) {
+            rssi = WiFi.RSSI(o);
+            selectedWiFi = n;
```

Comment:
This will always select the nearest WiFi, but the code will enter in a loop if the pass is wrong for that WiFi. Even if the second-nearest network has everything right.

In other words, if the nearest WiFi fails to establish a connection, then it should fallback to the second-nearest WiFi, and then to the third and go-on.

Reply:
I thought about that but decided it is best to leave AP open in such case to indicate incorrect set-up.
IMO people will use the functionality mostly when wifis are not available at the same time. But I may be wrong.

Reply:
IMO, we should always check every network, because sometimes AP doesn't work that fine as we would expect (as strange as it sounds, I have two cases of ESP32-S3 where I couldn't get the AP to work after a while. The WiFi/STA was working tho, just the AP).

The AP always must be the last resort for connection, it consumes a lot of power and heat.

Reply:
In such case just move a little. ;)

Joke aside, ESP has the capacity to be configured to connect to multiple defined APs.
I may implement that at some point.
https://docs.espressif.com/projects/arduino-esp32/en/latest/api/wifi.html#wifimulti

Reply:
In your first comment I thought that by leaving the AP open you meant about leaving the AP Mode open.

Instead of implementing ESP's wifimulti (which is only available in ESP32 AFAIK), we just should check the others networks if the first one doesn't works, as I said before.

Reply:
> Joke aside, ESP has the capacity to be configured to connect to multiple defined APs.  I may implement that at some point.

You guys just did.  I took a quick look at the internal implementation and it's [pretty close to the code in this PR.](https://github.com/espressif/arduino-esp32/blob/cceebb58f170c34a2356f0d22106dd6f0d141023/libraries/WiFi/src)/WiFiMulti.cpp#L119

It might be worth checking if it's easy to vendor WiFiMulti in to ESP8266 builds - I don't see anything glaringly obvious that would prevent it.